types/logger/logger.go

```diff
@@ -79,17 +83,17 @@ func (logf Logf) JSON(level int, recType string, v any) {
 }
 
 // FromContext extracts a log function from ctx.
+//
+// Deprecated: Use [LogfKey.Value] instead.
```

Comment:
Is there some reason _not_ to have a semantically-relevant helper? It's not clear to me we get much benefit from moving the thing the caller depends on from a simple helper function to a compound call on a value in the package. (I get that the helper is no longer _necessary_, but I think it reads better, esp. given that loggers are going to be used ~everywhere. But YMMV).

Reply:
Not a blocker, but I disagree with the weighting of that reasoning: Not that layout doesn't matter, but cognitive burden to the reader at the call site is IMO more important than, say, where the functions sort in a listing.

And we're still "polluting" the namespace, only now it's with a mutable variable (morally immutable, but as a reader when I see a variable reference I know it's a variable, and that makes me attend to whether that affects its use).

Reply:
For the callsite, we have:
|   | construction | extraction |
| -- | -- | -- |
| with `ctxkey` | `ctx := mypkg.ThingKey.WithValue(ctx, v)` | `v := mypkg.ThingKey.Value(ctx)` |
| user defined | `ctx := mypkg.WithThing(ctx, v)` | `v := mypkg.ThingValue(ctx)` |

Thoughts:

* The existence of the `ThingKey` namespace could be seen as additional cognitive burden.
* However, with `ctxkey`, we will always have a `WithValue(ctx, ...)` or `Value(ctx)` suffix, while with user-defined helpers, we're left with whether people follow a consistent convention (and we don't). I'd argue that consistency helps with cognitive burden.
* Using `ctxkey` is not always more typing as some helpers are named as `ContextWithThing` or `ThingFromContext`, which can be longer.

> it's a variable, and that makes me attend to whether that affects its use

Fair criticism. It'd be nice if I could make it a constant, but I don't know how. The alternative is to make it a argument-less function that just returns the `ctxkey.Key`, but that seems to buy us little.